adb/filesync_protocol.py

```diff
@@ -107,11 +109,24 @@ def Push(cls, connection, datafile, filename,
     cnxn = FileSyncConnection(connection, b'<2I')
     cnxn.Send(b'SEND', fileinfo)
 
+    bytes_written_len = 0
+    total_bytes = -1
+    if progress_callback and isinstance(datafile, file):
+      total_bytes = os.fstat(datafile.fileno()).st_size
+
     while True:
       data = datafile.read(MAX_PUSH_DATA)
-      if not data:
+      if data:
+        if progress_callback:
```

Comment:
you're calling the callback before actually sending data.

Also, can you push the progress-tracking into a small class? That way this loop remains simple and the `callback`/`total_bytes`/`bytes_written_len` is all tracked together